### PR TITLE
feat: add base classes for user-specified compaction

### PR DESCRIPTION
**Pull request by levy5307**
### What problem does this PR solve? 
RFC: https://github.com/apache/incubator-pegasus/pull/738

Add base classes for user-specified compaction. 
`compaction_operation` represents the compaction operation, such as update ttl/delete key
`compaction_filter_rule` represents the compaction rule to filter the keys which are stored in rocksdb.

##### Tests 

- Unit test
- No code actually.  I will add unit tests in later pull requests.

### CODE REVIEW

**Comment by acelyc111** [comment]
> ### What problem does this PR solve?
> Add base classes for user-specified compaction.
> `compaction_operation` represents the compaction operation, such as update ttl/delete key
> `compaction_filter_rule` represents the compaction rule to filter the keys which are stored in rocksdb.
> 
> ### Checklist
> ##### Tests
> * Unit test
> * No code actually.  I will add unit tests in later pull requests.

Better to add a proposal for this feature, even a Chinese doc.

File: src/server/compaction_filter_rule.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+#include <rocksdb/slice.h>
+
+namespace pegasus {
+namespace server {
+class compaction_filter_rule
+{
+public:
+    virtual ~compaction_filter_rule() = default;
+
+    // we can use `value_filed` to replace existing_value in the later,
```
Comment (acelyc111) [inline src/server/compaction_filter_rule.h:31]: Maybe it's a TODO?

File: src/server/compaction_operation.cpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace pegasus {
+namespace server {
+compaction_operation::~compaction_operation() = default;
+
+bool compaction_operation::all_rules_match(const std::string &hash_key,
+                                           const std::string &sort_key,
+                                           const rocksdb::Slice &existing_value) const
+{
+    if (rules.size() == 0) {
```
Comment (acelyc111) [inline src/server/compaction_operation.cpp:30]:
```suggestion
    if (rules.empty()) {
```

File: src/server/compaction_operation.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+#include <memory>
+#include <vector>
+#include "compaction_filter_rule.h"
+
+namespace pegasus {
+namespace server {
+
+typedef std::vector<std::unique_ptr<compaction_filter_rule>> filter_rules;
+class compaction_operation
```
Comment (acelyc111) [inline src/server/compaction_operation.h:30]: Would you please to add a sample description for this class?

**Comment by hycdong** [comment]
I agree with @acelyc111 , expecting your proposal and comments.

File: src/server/compaction_filter_rule.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+namespace server {
+/** compaction_filter_rule represents the compaction rule to filter the keys which are stored in
+ * rocksdb. */
+class compaction_filter_rule
+{
+public:
+    virtual ~compaction_filter_rule() = default;
+
+    // TODO(zhaoliwei): we can use `value_filed` to replace existing_value in the later,
+    // after the refactor of value schema
```
Comment (hycdong) [inline src/server/compaction_filter_rule.h:34]: The refactor of value schema has already been merged, you can update it in next pull request :-)

**Review by hycdong — APPROVED** [review]

File: src/server/compaction_filter_rule.h (hunk 4 above)
Comment (levy5307) [reply src/server/compaction_filter_rule.h:34]: Yes I will refactor it in the later

**Review by acelyc111 — APPROVED** [review]

**Pull request merged**